File: lib/core/units.dart
enum UnitSystem { metric, imperial }

extension UnitSystemX on UnitSystem {
  String get id => switch (this) {
        UnitSystem.metric => 'metric',
        UnitSystem.imperial => 'imperial',
      };

  String get weightLabel => this == UnitSystem.metric ? 'kg' : 'lb';
  String get lengthLabel => this == UnitSystem.metric ? 'cm' : 'in';

  static UnitSystem fromId(String? id) =>
      id == 'imperial' ? UnitSystem.imperial : UnitSystem.metric;
}

/// Conversions between metric storage and the user's display unit.
class UnitConverter {
  const UnitConverter(this.system);
  final UnitSystem system;

  static const double _kgPerLb = 0.45359237;
  static const double _cmPerIn = 2.54;

  /// kg (storage) -> display weight
  double weightFromKg(double? kg) {
    if (kg == null) return 0;
    return system == UnitSystem.metric ? kg : kg / _kgPerLb;
  }

  /// display weight -> kg (storage)
  double weightToKg(double value) =>
      system == UnitSystem.metric ? value : value * _kgPerLb;

  /// cm (storage) -> display length
  double lengthFromCm(double? cm) {
    if (cm == null) return 0;
    return system == UnitSystem.metric ? cm : cm / _cmPerIn;
  }

  /// display length -> cm (storage)
  double lengthToCm(double value) =>
      system == UnitSystem.metric ? value : value * _cmPerIn;

  String formatWeight(double? kg, {int digits = 1}) {
    if (kg == null) return '—';
    return '${weightFromKg(kg).toStringAsFixed(digits)} ${system.weightLabel}';
  }

  String formatLength(double? cm, {int digits = 1}) {
    if (cm == null) return '—';
    return '${lengthFromCm(cm).toStringAsFixed(digits)} ${system.lengthLabel}';
  }
}

File: src/core/units.ts
export type UnitSystem = 'metric' | 'imperial';

const KG_PER_LB = 0.45359237;
const CM_PER_IN = 2.54;

export const weightLabel = (u: UnitSystem) => (u === 'metric' ? 'kg' : 'lb');
export const lengthLabel = (u: UnitSystem) => (u === 'metric' ? 'cm' : 'in');

export const weightFromKg = (kg: number | null | undefined, u: UnitSystem) =>
  kg == null ? null : u === 'metric' ? kg : kg / KG_PER_LB;

export const weightToKg = (value: number, u: UnitSystem) =>
  u === 'metric' ? value : value * KG_PER_LB;

export const lengthFromCm = (cm: number | null | undefined, u: UnitSystem) =>
  cm == null ? null : u === 'metric' ? cm : cm / CM_PER_IN;

export const lengthToCm = (value: number, u: UnitSystem) =>
  u === 'metric' ? value : value * CM_PER_IN;

export const formatWeight = (
  kg: number | null | undefined,
  u: UnitSystem,
  digits = 1,
) => {
  const v = weightFromKg(kg, u);
  return v == null ? '—' : `${v.toFixed(digits)} ${weightLabel(u)}`;
};

export const formatLength = (
  cm: number | null | undefined,
  u: UnitSystem,
  digits = 1,
) => {
  const v = lengthFromCm(cm, u);
  return v == null ? '—' : `${v.toFixed(digits)} ${lengthLabel(u)}`;
};
